### PR TITLE
Implement YAML roundtripping, comment preserving edits

### LINKED ISSUE #423: Support full fidelity round-tripping of Pulumi.[yaml|json]
When writing information to the `Pulumi.yaml` file, we end up serializing back the entire structure and that serialization works by serializing each field in the `pack.Package` in the order it was defined in the go object.

In the case of `YAML` any comments would also be removed from the file.

This means if we had a Pulumi.yaml file that looked like this:

```
runtime: nodejs
name: hello-world
```
If we read and save it it would look like this:

```
name: hello-world
runtime: nodejs
```

Ideally, we would have parsers that would turn these files not into raw go objects but some DOM that we could round trip with full fidelity.

### LINKED ISSUE #5235: Pulumi Should retain comments in Stack Config
When running Pulumi commands like `pulumi up` or `pulumi preview`  when using Azure KeyVault as it's Secrets Provider for the selected stac, comments in the relevant config file are removed. It would be useful to retain these comments as config files could be edited by hand and checked into source control with these comments to provide information to other users about the reasoning for the config value. When not using the `--secrets-provider` flag. Running `pulumi` commands retain comments in the config as expected.

Example stack config file `pulumi.<stack>.yaml`:
```
# Top level comment
secretsprovider: REDACTED
encryptedkey: REDACTED
config:
  # Another Comment
  MyProject:GroupIds: 
  - 11111111-1111-1111-1111-111111111111 # Group one
  - 11111111-1111-1111-1111-111111111112 # Group 2
```

## Expected Behavior

Comments should be retained in the config file as these could contain useful information that refers to config values

## Current Behavior

All comments are removed from config

## Steps to Reproduce

1. Create a new project `pulumi new csharp`
2. Initialise the stack with Azure KeyVault as the secrets provider
```
pulumi stack init dev --secrets-provider=azurekeyvault://<Vault Name>.vault.azure.net/keys/<Key Name>
```
3. Open the `pulumi.dev.yaml` file and add a comment
4. Run `pulumi up` and comments will be removed

### PR DESCRIPTION
**Pull request by AaronFriel**
By caching a copy of the raw bytes of the original loaded document - project, policy project, or stack - on save we:

* Unmarshal the original bytes to a YAML AST
* Recursively edit nodes of the document to match the values of the value we're saving, preserving trivia
* Marshal the modified AST to []byte

Closes #423, as we don't support comments in JSON presently (not using "JSON5")

Closes #5235

Simplifies work done in #10797 and #10437 which used the previous `yamlutil` package to edit AST nodes directly.

# Automation API

Integration with automation API requires that operations either perform a "read-modify-write" in a single operation, or use a side channel to allow subsequent "read" (select stack?) and "write" operations to keep state of the original file.

However I don't expect this to be needed. We don't have a reason to believe automation API users maintain comments in particular stack files, and doing so is already unsupported in automation scenarios. The Pulumi Service does not persist raw config files, thus `pulumi config refresh` and `CreateOrSelectStack` create config files without comments.

n.b.: If you test this, due to the new behavior of this PR, `pulumi config refresh` will persist comments if there is an existing config file. I would expect automation API using local file workspaces to behave similarly. When using these tools to create a config file from the last deployment's definition, the file is created restored without comments.

# Tradeoffs

I believe that this closes #423 in spirit, but doesn't strictly implement round-tripping. There are a couple reasons for this.

## Marshaller interface and round-tripping simplification decisions

We support unmarshaling "simplified" YAML, including allowing:

```yaml
runtime: yaml # this unmarshals to a struct, not a scalar value
config:
  someKey: "someScalar" # likewise
```

These simplifications make it much more difficult, the `Marshal` operation would _also_ need context of decisions made when unmarshaling to ensure a bijection. Otherwise we may save one of these as the other:

```yaml
runtime:
  name: yaml
```

```yaml
runtime: yaml
```

This could be considered a feature, as it enforces a style guide - the decisions implemented in our custom `Marshaller` implementations become the style enforced on save.

That said, I believe the new `Edit` API while not round-trippable, does reach a fixed point after one save through the new API. Once saved with this API, subsequent loads & saves without changes are idempotent.

## Indentation

The YAML library uses a global indentation when marshaling values to YAML, as opposed to indentation being a property of AST nodes. This means we may lose other indentation.

### CODE REVIEW

File: sdk/go/common/util/yamlutil/edit.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+	var b bytes.Buffer
+	yamlEncoder := yaml.NewEncoder(&b)
+	yamlEncoder.SetIndent(2)
+	err = yamlEncoder.Encode(&newValue)
+	if err != nil {
+		return nil, fmt.Errorf("error editing value %#v: %w", newValue, err)
+	}
+
+	return b.Bytes(), nil
```
Comment (iwahbe) [inline sdk/go/common/util/yamlutil/edit.go:47]: If there was a way to avoid repeating this in `sdk/go/common/encoding/marshal.go`, that would be great.

**Review by iwahbe — APPROVED** [review]
Would it be possible to piggyback an integration test for `pulumi config set` for this change? 

Looks good too me.

File: sdk/go/common/util/yamlutil/edit.go (hunk 1 above)
Comment (AaronFriel) [reply sdk/go/common/util/yamlutil/edit.go:47]: Yeah I think so - we could modify an integration test to start with a file that has some config and comments.

**Comment by AaronFriel** [comment]
@iwahbe re: testing. Yeah I think so - we could modify an integration test to start with a file that has some config and comments. I'll do that.

**Comment by AaronFriel** [comment]
@Frassle @justinvp @iwahbe Added tests here: https://github.com/pulumi/pulumi/blob/friel/yaml-roundtrip/tests/roundtrip_test.go#L1-L182

I'd like the first test to use `pulumi config set --project` when that is implemented. For now, it loads and saves the project in the same manner as the CLI does.

File: sdk/go/common/util/yamlutil/edit.go (hunk 1 above)
Comment (AaronFriel) [reply sdk/go/common/util/yamlutil/edit.go]: Exported a two space indendation `YamlEncode` and refactored.

**Comment by AaronFriel** [comment]
bors merge

**Comment by AaronFriel** [comment]
bors retry

**Pull request merged**